php/libraries/SinglePointLogin.class.inc

```diff
@@ -390,6 +404,89 @@ class SinglePointLogin
+     * attempts.
+     *
+     * @param string $username The user account to check.
+     *
+     * @return bool
+     */
+    function accountLocked(string $username): bool
+    {
+        /* A user is locked out if they have made X attempts in the last Y
+         * minutes or X' attempts in the last Y' minutes.
+         * The reason for using two time windows is to provide some flexibility:
+         * legitimate locked-out users to try again in a short time while also
+         * preventing attackers from simply trying another X login attempts
+         * after Y minutes.
+         */
+
+        // These represent X and X' above, but with more helpful names.
+        $shortWindowInMinutes = 15;
+        $longWindowInMinutes  = 60;
+        /* These represent Y and Y' above, but with more helpful names.
+         * These values represent the maximum umber of attempts a user may make
+         * to login before being locked out.
+         */
+        $shortLockoutThreshold = 10;
+        $longLockoutThreshold  = 15;
+
+        $shortFailCount = $this->countFailedLogins(
+            $username,
+            $shortWindowInMinutes
+        );
+        $longFailCount  = $this->countFailedLogins(
+            $username,
+            $longWindowInMinutes
+        );
+        return $shortFailCount > $shortLockoutThreshold
+            || $longFailCount > $longLockoutThreshold;
+    }
+
+    /**
+     * Gives the number of failed login attempts that occurred in the last
+     * $timeWindowInMinutes minutes. Used to determine whether a user's login
+     * request should be denied as a security precaution.
+     *
+     * @param string $username            The user account to check.
+     * @param int    $timeWindowInMinutes This value is subtracted from the
+     *                          current timestamp to find relevant failed
+     *                          login attempts. Only more recent failed
+     *                          attempts are considered relevant here.
+     *
+     * @return int The number of failed attempts in the time window.
+     */
+    function countFailedLogins(
+        string $username,
+        int $timeWindowInMinutes
+    ): int {
+        $factory = NDB_Factory::singleton();
+        $db      = $factory->database();
+        $query   = "SELECT COUNT(loginhistoryID)" .
+            "FROM user_login_history " .
+            "WHERE userID = :username " .
+            "AND Success = 'N' " .
+            "AND Login_timestamp > ( " .
+            "SELECT IF( MAX(Login_timestamp) > DATE_SUB(NOW(), ".
+            "INTERVAL :timeWindowInMinutes MINUTE), " .
+                "MAX(Login_timestamp), DATE_SUB(NOW(), " .
+                "INTERVAL :timeWindowInMinutes MINUTE) " .
+            ") " .
+            "FROM user_login_history " .
+            "WHERE userID = :username " .
+            "AND Success = 'Y'" .
+        ");";
+        $result  = $db->pselect(
```

Comment:
When you make the same query with different parameters, it would be more efficient to use the same prepared query twice.

Reply:
Oops this shouldn't be here.

Reply:
@xlecours There is only 1 query here, with a nested but different query build-in.
@johnsaigle quey is fine

Reply:
@xlecours Could you follow-up here or else dismiss your review?

Reply:
The `countFailedLogins` function is called twice. I think this is exacly why it as been put in a function. You could have prepare the query and then call execute twice with the different parameters. 
```

Reply:
This doesn't seem to be addressed

Reply:
@driusan @xlecours I updated the code after this comment. The difference can be seen between the commits on November 12 and 13. As I stated below, please let me know whether this version looks better to you than before. I'm not totally sure I understood the problem but I addressed it to my understanding.